lib/linked_list.rb

```diff
@@ -18,71 +18,148 @@ def initialize
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+        new_node = Node.new(value)
+      if @head.nil?
+        @head = new_node
+      else 
+        new_node.next = @head
+        @head = new_node
+      end
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      if @head.nil?
+        return false
+      end
+      current = @head
+      until current.data == value || current.next.nil?
+        current = current.next
+      end
+      return current.data == value
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      max = @head.data
+      until current.next.nil?
+        current = current.next
+        if current.data > max
+          max = current.data
+        end
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_min
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      min = @head.data
+      until current.next.nil?
+        current = current.next
+        if current.data < min
+          min = current.data
+        end
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      if @head.nil?
+        return 0
+      end
+      count = 1
+      current = @head
+      until current.next.nil?
+        current = current.next
+        count += 1
+      end
+      return count

```

Comment:
This can be simplified a bit.

```suggestion
      count = 0
      current = @head
      until current.nil? 
        current = current.next
        count += 1
      end
      return count
      
```